optimism/src/mips/proof.rs

```diff
@@ -24,6 +25,70 @@ pub struct WitnessColumns<G> {
     pub error: G,
 }
 
+impl<G> IntoParallelIterator for WitnessColumns<G>
+where
+    Vec<G>: IntoParallelIterator,
+{
+    type Iter = <Vec<G> as IntoParallelIterator>::Iter;
+    type Item = <Vec<G> as IntoParallelIterator>::Item;
+
+    fn into_par_iter(self) -> Self::Iter {
+        let mut iter_contents = Vec::with_capacity(crate::mips::witness::SCRATCH_SIZE + 2);
+        iter_contents.extend(self.scratch);
+        iter_contents.push(self.instruction_counter);
+        iter_contents.push(self.error);
+        iter_contents.into_par_iter()
+    }
+}
+
+impl<G: Send + std::fmt::Debug> FromParallelIterator<G> for WitnessColumns<G> {
+    fn from_par_iter<I>(par_iter: I) -> Self
+    where
+        I: IntoParallelIterator<Item = G>,
+    {
+        let mut iter_contents = par_iter.into_par_iter().collect::<Vec<_>>();
+        let error = iter_contents.pop().unwrap();
+        let instruction_counter = iter_contents.pop().unwrap();
```

Comment:
Silly question whose answer should be "yes", but asking just in case: does push and pop for vectors behave in "the expected way"? Meaning, that the elements in the resulting vector being popped occur in the same order as the ones being pushed 🤔

Reply:
Yes :)